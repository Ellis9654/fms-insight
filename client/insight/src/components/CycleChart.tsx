/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
import * as React from 'react';
import * as im from 'immutable';
import { format } from 'date-fns';
import { MarkSeries,
         XAxis,
         YAxis,
         Hint,
         FlexibleWidthXYPlot,
         VerticalGridLines,
         HorizontalGridLines,
         DiscreteColorLegend
       } from 'react-vis';
import Card, { CardHeader, CardContent } from 'material-ui/Card';
import * as numerable from 'numeral';
import Select from 'material-ui/Select';
import { MenuItem } from 'material-ui/Menu';

export interface CycleChartPoint {
  readonly x: Date;
  readonly y: number;
}

export interface CycleChartProps {
  points: im.Map<string, ReadonlyArray<CycleChartPoint>>;
  series_label: string;
}

interface CycleChartTooltip {
  readonly x: Date;
  readonly y: number;
  readonly series: string;
}

interface CycleChartState {
  tooltip?: CycleChartTooltip;
  disabled_series: { [key: string]: boolean };
}

export class CycleChart extends React.PureComponent<CycleChartProps, CycleChartState> {
  state = {
    tooltip: undefined,
    disabled_series: {}
  } as CycleChartState;

  setClosestPoint = (series: string) => (point: CycleChartPoint) => {
    if (this.state.tooltip === undefined) {
      this.setState({tooltip: {...point, series: series}});
    } else {
      this.setState({tooltip: undefined});
    }
  }

  clearTooltip = () => {
    this.setState({tooltip: undefined});
  }

  toggleSeries = (series: {title: string}) => {
    const newState = !!!this.state.disabled_series[series.title];
    this.setState({
      disabled_series: {...this.state.disabled_series,
        [series.title]: newState
      }
    });
  }

  formatHint = (tip: CycleChartTooltip) => {
    return [
      {title: 'Time', value: format(tip.x, 'MMM D, YYYY, H:mm a')},
      {title: this.props.series_label, value: tip.series},
      {title: 'Cycle Time', value: numerable(tip.y).format('0.0') + " minutes"},
    ];
  }

  render() {
    const seriesNames =
      this.props.points.toSeq()
      .sortBy((points, s) => s);
    return (
      <div>
        <FlexibleWidthXYPlot
            height={window.innerHeight - 200}
            xType="time"
            margin={{bottom: 50}}
            onMouseLeave={this.clearTooltip}
        >
          <VerticalGridLines/>
          <HorizontalGridLines/>
          <XAxis tickLabelAngle={-45}/>
          <YAxis/>
          {
            seriesNames.map((points, series) =>
              <MarkSeries
                key={series}
                data={points}
                onValueClick={this.setClosestPoint(series)}
                {...(this.state.disabled_series[series] ? {opacity: 0.2} : null)}
              />
            ).toIndexedSeq()
          }
          {
            this.state.tooltip === undefined ? undefined :
              <Hint value={this.state.tooltip} format={this.formatHint}/>
          }
        </FlexibleWidthXYPlot>

        <div style={{textAlign: 'center'}}>
          <DiscreteColorLegend
            orientation="horizontal"
            items={
              seriesNames.keySeq().map(s =>
                ({title: s, disabled: this.state.disabled_series[s]})
              ).toArray()
            }
            onItemClick={this.toggleSeries}
          />
        </div>
      </div>
    );
  }
}

export interface SelectableCycleChartProps {
  points: im.Map<string, im.Map<string, ReadonlyArray<CycleChartPoint>>>;
  select_label: string;
  series_label: string;
  card_label: string;
  icon: JSX.Element;
}

export class SelectableCycleChart extends React.PureComponent<SelectableCycleChartProps, {selected: string}> {
  state = {selected: ""};

  setSelected = (selected: string) => {
    this.setState({selected});
  }

  render() {
    return (
      <Card>
        <CardHeader
          title={
            <div style={{display: 'flex', flexWrap: 'wrap', alignItems: 'center'}}>
              {this.props.icon}
              <div style={{marginLeft: '10px', marginRight: '3em'}}>
                {this.props.card_label}
              </div>
              <div style={{flexGrow: 1}}/>
              <Select
                autoWidth
                displayEmpty
                value={this.state.selected}
                onChange={e => this.setSelected(e.target.value)}
              >
                {
                  this.state.selected !== "" ? undefined :
                    <MenuItem key={0} value=""><em>Select {this.props.select_label}</em></MenuItem>
                }
                {
                  this.props.points.keySeq().sort().map(n =>
                    <MenuItem key={n} value={n}>{n}</MenuItem>
                  )
                }
              </Select>
            </div>}
        />
        <CardContent>
          <CycleChart
            points={this.props.points.get(this.state.selected, im.Map())}
            series_label={this.props.series_label}
          />
        </CardContent>
      </Card>
    );
  }
}