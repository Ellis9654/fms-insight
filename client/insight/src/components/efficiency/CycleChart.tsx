/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
import * as React from 'react';
import * as im from 'immutable';
import { format, addDays } from 'date-fns';
import { MarkSeries,
         XAxis,
         YAxis,
         Hint,
         FlexibleWidthXYPlot,
         VerticalGridLines,
         HorizontalGridLines,
         DiscreteColorLegend
       } from 'react-vis';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import CardHeader from '@material-ui/core/CardHeader';
import * as numerable from 'numeral';
import Select from '@material-ui/core/Select';
import MenuItem from '@material-ui/core/MenuItem';
import { PartIdenticon } from '../station-monitor/Material';

export interface CycleChartPoint {
  readonly x: Date;
  readonly y: number;
}

export interface CycleChartProps {
  readonly points: im.Map<string, ReadonlyArray<CycleChartPoint>>;
  readonly series_label: string;
  readonly default_date_range?: Date[];
}

interface CycleChartTooltip {
  readonly x: Date;
  readonly y: number;
  readonly series: string;
}

interface CycleChartState {
  readonly tooltip?: CycleChartTooltip;
  readonly disabled_series: { [key: string]: boolean };
}

function memoize<A, R>(f: (x: A) => R): ((x: A) => R) {
  let memo = new Map<A, R>();
  return x => {
    let ret = memo.get(x);
    if (!ret) {
      ret = f(x);
      memo.set(x, ret);
    }
    return ret;
  };
}

export class CycleChart extends React.PureComponent<CycleChartProps, CycleChartState> {
  state = {
    tooltip: undefined,
    disabled_series: {}
  } as CycleChartState;

  // memoize on the series name, since the function from CycleChartPoint => void is
  // passed as a prop to the chart, and reusing the same function keeps the props
  // unchanged so PureComponent can avoid a re-render
  setTooltip = memoize((series: string) => (point: CycleChartPoint) => {
    if (this.state.tooltip === undefined) {
      this.setState({tooltip: {...point, series: series}});
    } else {
      this.setState({tooltip: undefined});
    }
  });

  clearTooltip = () => {
    this.setState({tooltip: undefined});
  }

  toggleSeries = (series: {title: string}) => {
    const newState = !!!this.state.disabled_series[series.title];
    this.setState({
      disabled_series: {...this.state.disabled_series,
        [series.title]: newState
      }
    });
  }

  formatHint = (tip: CycleChartTooltip) => {
    return [
      {title: 'Time', value: format(tip.x, 'MMM D, YYYY, H:mm a')},
      {title: this.props.series_label, value: tip.series},
      {title: 'Cycle Time', value: numerable(tip.y).format('0.0') + " minutes"},
    ];
  }

  render() {
    const seriesNames =
      this.props.points.toSeq()
      .sortBy((points, s) => s);

    let dateRange = this.props.default_date_range;
    if (dateRange === undefined) {
      const now = new Date();
      const oneMonthAgo = addDays(now, -30);
      dateRange = [now, oneMonthAgo];
    }

    return (
      <div>
        <FlexibleWidthXYPlot
            height={window.innerHeight - 200}
            xType="time"
            margin={{bottom: 50}}
            onMouseLeave={this.clearTooltip}
            dontCheckIfEmpty
            xDomain={this.props.points.isEmpty() ? dateRange : undefined}
            yDomain={this.props.points.isEmpty() ? [0, 60] : undefined}
        >
          <VerticalGridLines/>
          <HorizontalGridLines/>
          <XAxis tickLabelAngle={-45}/>
          <YAxis/>
          {
            seriesNames.map((points, series) =>
              <MarkSeries
                key={series}
                data={points}
                onValueClick={this.state.disabled_series[series] ? undefined : this.setTooltip(series)}
                {...(this.state.disabled_series[series] ? {opacity: 0.2} : null)}
              />
            ).toIndexedSeq()
          }
          {
            this.state.tooltip === undefined ? undefined :
              <Hint value={this.state.tooltip} format={this.formatHint}/>
          }
        </FlexibleWidthXYPlot>

        <div style={{textAlign: 'center'}}>
          <DiscreteColorLegend
            orientation="horizontal"
            items={
              seriesNames.keySeq().map(s =>
                ({title: s, disabled: this.state.disabled_series[s]})
              ).toArray()
            }
            onItemClick={this.toggleSeries}
          />
        </div>
      </div>
    );
  }
}

export interface SelectableCycleChartProps {
  readonly points: im.Map<string, im.Map<string, ReadonlyArray<CycleChartPoint>>>;
  readonly select_label: string;
  readonly series_label: string;
  readonly card_label: string;
  readonly icon: JSX.Element;
  readonly default_date_range?: Date[];
  readonly selected?: string;
  readonly useIdenticon?: boolean;
  readonly setSelected: (s: string) => void;
}

export function SelectableCycleChart(props: SelectableCycleChartProps) {
  let validValue = props.selected !== undefined && props.points.has(props.selected);
  function stripAfterDash(s: string): string {
    const idx = s.indexOf('-');
    if (idx >= 0) {
      return s.substring(0, idx);
    } else {
      return s;
    }
  }
  return (
    <Card raised>
      <CardHeader
        title={
          <div style={{display: 'flex', flexWrap: 'wrap', alignItems: 'center'}}>
            {props.icon}
            <div style={{marginLeft: '10px', marginRight: '3em'}}>
              {props.card_label}
            </div>
            <div style={{flexGrow: 1}}/>
            <Select
              autoWidth
              displayEmpty
              value={validValue ? props.selected : ""}
              onChange={e => props.setSelected(e.target.value)}
            >
              {
                validValue ? undefined :
                  <MenuItem key={0} value=""><em>Select {props.select_label}</em></MenuItem>
              }
              {
                props.points.keySeq().sort().map(n =>
                  <MenuItem key={n} value={n}>
                    <div style={{display: "flex", alignItems: "center"}}>
                      <PartIdenticon part={stripAfterDash(n)} size={30}/>
                      <span style={{marginRight: '1em'}}>{n}</span>
                    </div>
                  </MenuItem>
                )
              }
            </Select>
          </div>}
      />
      <CardContent>
        <CycleChart
          points={props.points.get(props.selected || "", im.Map())}
          series_label={props.series_label}
          default_date_range={props.default_date_range}
        />
      </CardContent>
    </Card>
  );
  }