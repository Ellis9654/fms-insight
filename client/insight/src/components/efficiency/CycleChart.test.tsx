/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import * as React from 'react';
import * as im from 'immutable';
import { shallow } from 'enzyme';
import WorkIcon from '@material-ui/icons/Work';

import { CycleChart, CycleChartProps, SelectableCycleChartProps, SelectableCycleChart } from './CycleChart';

it('displays a cycle chart', () => {
  const props: CycleChartProps = {
    points:
      im.Map({
        "aaa": [{x: new Date(Date.UTC(2018, 1, 1)), y: 15}, {x: new Date(Date.UTC(2018, 1, 2)), y: 66}],
        "bbb": [{x: new Date(Date.UTC(2018, 2, 2)), y: 15}, {x: new Date(Date.UTC(2018, 2, 3)), y: 55}],
        "ccc": [{x: new Date(Date.UTC(2018, 3, 3)), y: 15}],
      }),
    series_label: "serieslabel"
  };

  const val = shallow(<CycleChart {...props}/>);
  expect(val).toMatchSnapshot('cycle chart');
});

it("displays an empty chart", () => {
  const props: CycleChartProps = {
    points: im.Map(),
    series_label: "serieslabel",
    default_date_range: [new Date(Date.UTC(2018, 2, 5)), new Date(Date.UTC(2018, 2, 10))],
  };

  const val = shallow(<CycleChart {...props}/>);
  expect(val).toMatchSnapshot('empty cycle chart');
});

const selectProps: SelectableCycleChartProps = {
    points:
      im.Map({
        "111": im.Map({
            "aaa": [{x: new Date(Date.UTC(2018, 1, 1)), y: 15}, {x: new Date(Date.UTC(2018, 1, 2)), y: 66}],
            "bbb": [{x: new Date(Date.UTC(2018, 2, 2)), y: 53}, {x: new Date(Date.UTC(2018, 2, 3)), y: 55}],
            "ccc": [{x: new Date(Date.UTC(2018, 3, 3)), y: 6}],
          }),
        "222": im.Map({
            "aaa": [{x: new Date(Date.UTC(2018, 4, 4)), y: 62}, {x: new Date(Date.UTC(2018, 4, 45)), y: 512}],
          }),
      }),
    series_label: "serieslabel",
    select_label: "sellabel",
    card_label: 'cardheader',
    icon: <WorkIcon/>,
    selected: undefined,
    setSelected: jest.fn()
  };

it('displays a selectable cycle chart with nothing selected', () => {
  const props = {...selectProps, selected: 'selectedval'};
  const val = shallow(<SelectableCycleChart {...props}/>);
  expect(val).toMatchSnapshot('selectable cycle chart with nothing selected');
});

it('displays a selectable cycle chart with something selected', () => {
  const props = {...selectProps, selected: '111'};
  const val = shallow(<SelectableCycleChart {...props}/>);
  expect(val).toMatchSnapshot('selectable cycle chart with something selected');
});
