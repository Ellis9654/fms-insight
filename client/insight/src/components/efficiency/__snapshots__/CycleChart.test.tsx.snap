// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays a cycle chart: cycle chart 1`] = `
<div>
  <FlexibleXYPlot
    dontCheckIfEmpty={true}
    height={568}
    margin={
      Object {
        "bottom": 50,
      }
    }
    onMouseLeave={[Function]}
    xType="time"
  >
    <VerticalGridLines
      attr="x"
      direction="vertical"
    />
    <HorizontalGridLines
      attr="y"
      direction="horizontal"
    />
    <XAxis
      attr="x"
      attrAxis="y"
      orientation="bottom"
      tickLabelAngle={-45}
    />
    <YAxis
      attr="y"
      attrAxis="x"
      orientation="left"
    />
    <MarkSeries
      data={
        Array [
          Object {
            "x": 2018-02-01T00:00:00.000Z,
            "y": 15,
          },
          Object {
            "x": 2018-02-02T00:00:00.000Z,
            "y": 66,
          },
        ]
      }
      getNull={[Function]}
      key="aaa"
      onValueClick={[Function]}
    />
    <MarkSeries
      data={
        Array [
          Object {
            "x": 2018-03-02T00:00:00.000Z,
            "y": 15,
          },
          Object {
            "x": 2018-03-03T00:00:00.000Z,
            "y": 55,
          },
        ]
      }
      getNull={[Function]}
      key="bbb"
      onValueClick={[Function]}
    />
    <MarkSeries
      data={
        Array [
          Object {
            "x": 2018-04-03T00:00:00.000Z,
            "y": 15,
          },
        ]
      }
      getNull={[Function]}
      key="ccc"
      onValueClick={[Function]}
    />
  </FlexibleXYPlot>
  <div
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    <DiscreteColorLegendItem
      className=""
      colors={
        Array [
          "#12939A",
          "#79C7E3",
          "#1A3177",
          "#FF9833",
          "#EF5D28",
        ]
      }
      items={
        Array [
          Object {
            "disabled": undefined,
            "title": "aaa",
          },
          Object {
            "disabled": undefined,
            "title": "bbb",
          },
          Object {
            "disabled": undefined,
            "title": "ccc",
          },
        ]
      }
      onItemClick={[Function]}
      orientation="horizontal"
    />
  </div>
</div>
`;

exports[`displays a selectable cycle chart with nothing selected: selectable cycle chart with nothing selected 1`] = `
<WithStyles(Card)
  raised={true}
>
  <WithStyles(CardHeader)
    title={
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexWrap": "wrap",
          }
        }
      >
        <pure(Work) />
        <div
          style={
            Object {
              "marginLeft": "10px",
              "marginRight": "3em",
            }
          }
        >
          cardheader
        </div>
        <div
          style={
            Object {
              "flexGrow": 1,
            }
          }
        />
        <WithStyles(Select)
          autoWidth={true}
          displayEmpty={true}
          onChange={[Function]}
          value=""
        >
          <WithStyles(MenuItem)
            value=""
          >
            <em>
              Select 
              sellabel
            </em>
          </WithStyles(MenuItem)>
          Immutable.Seq […]
        </WithStyles(Select)>
      </div>
    }
  />
  <WithStyles(CardContent)>
    <CycleChart
      points={Immutable.Map {}}
      series_label="serieslabel"
    />
  </WithStyles(CardContent)>
</WithStyles(Card)>
`;

exports[`displays a selectable cycle chart with something selected: selectable cycle chart with something selected 1`] = `
<WithStyles(Card)
  raised={true}
>
  <WithStyles(CardHeader)
    title={
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flexWrap": "wrap",
          }
        }
      >
        <pure(Work) />
        <div
          style={
            Object {
              "marginLeft": "10px",
              "marginRight": "3em",
            }
          }
        >
          cardheader
        </div>
        <div
          style={
            Object {
              "flexGrow": 1,
            }
          }
        />
        <WithStyles(Select)
          autoWidth={true}
          displayEmpty={true}
          onChange={[Function]}
          value="111"
        >
          Immutable.Seq […]
        </WithStyles(Select)>
      </div>
    }
  />
  <WithStyles(CardContent)>
    <CycleChart
      points={
        Immutable.Map {
          "aaa": Array [
            Object {
              "x": 2018-02-01T00:00:00.000Z,
              "y": 15,
            },
            Object {
              "x": 2018-02-02T00:00:00.000Z,
              "y": 66,
            },
          ],
          "bbb": Array [
            Object {
              "x": 2018-03-02T00:00:00.000Z,
              "y": 53,
            },
            Object {
              "x": 2018-03-03T00:00:00.000Z,
              "y": 55,
            },
          ],
          "ccc": Array [
            Object {
              "x": 2018-04-03T00:00:00.000Z,
              "y": 6,
            },
          ],
        }
      }
      series_label="serieslabel"
    />
  </WithStyles(CardContent)>
</WithStyles(Card)>
`;

exports[`displays an empty chart: empty cycle chart 1`] = `
<div>
  <FlexibleXYPlot
    dontCheckIfEmpty={true}
    height={568}
    margin={
      Object {
        "bottom": 50,
      }
    }
    onMouseLeave={[Function]}
    xDomain={
      Array [
        2018-03-05T00:00:00.000Z,
        2018-03-10T00:00:00.000Z,
      ]
    }
    xType="time"
    yDomain={
      Array [
        0,
        60,
      ]
    }
  >
    <VerticalGridLines
      attr="x"
      direction="vertical"
    />
    <HorizontalGridLines
      attr="y"
      direction="horizontal"
    />
    <XAxis
      attr="x"
      attrAxis="y"
      orientation="bottom"
      tickLabelAngle={-45}
    />
    <YAxis
      attr="y"
      attrAxis="x"
      orientation="left"
    />
  </FlexibleXYPlot>
  <div
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    <DiscreteColorLegendItem
      className=""
      colors={
        Array [
          "#12939A",
          "#79C7E3",
          "#1A3177",
          "#FF9833",
          "#EF5D28",
        ]
      }
      items={Array []}
      onItemClick={[Function]}
      orientation="horizontal"
    />
  </div>
</div>
`;
